### PR TITLE
Enhance gateway to support secure web proxy

### PR DESCRIPTION
**Pull request by felipegc**
fixes https://github.com/hashicorp/terraform-provider-google/issues/14324 

Adding necessary fields and treatments in network service gateway for secure web proxy type.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
networkservices: added necessary fields to `google_network_services_gateway` to make it compatible with secure web proxy (beta)
```

```release-note:bug
networkservices: fixed a bug where modifying non-updatable fields `scope` in `google_network_services_gateway` would fail with API errors; now updating them will recreate the resource (beta)
```

### CODE REVIEW

**Comment by shuyama1** [comment]
@felipegc Could you share the API doc that you followed adding this resource if there's any. Thanks!

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (shuyama1) [inline mmv1/templates/terraform/constants/network_services_gateway.go.erb:43]: Only nit since we return `true` if the deleted one is the last gateway of type `SECURE_WEB_GATEWAY`: 
```suggestion
// Checks if the given list of gateways contains no other gateways of type SECURE_WEB_GATEWAY.
```

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (shuyama1) [inline mmv1/templates/terraform/constants/network_services_gateway.go.erb:18]: nit: spacing: should be tabs instead of spaces

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (shuyama1) [inline mmv1/templates/terraform/constants/network_services_gateway.go.erb:25]: we have a recent rework of `SendRequest` (https://github.com/GoogleCloudPlatform/magic-modules/pull/793). Please follow the new structure. Apologize for the extra work needed.

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (shuyama1) [inline mmv1/templates/terraform/constants/network_services_gateway.go.erb:82]: same here for `SendRequest `

File: mmv1/products/networkservices/Gateway.yaml
Comment (shuyama1) [inline mmv1/products/networkservices/Gateway.yaml:204]: Do `addresses`, `subnetwork`, `network` support in-place update? If so, can we add an update test? Otherwise, we need to add `immutable`.

File: mmv1/third_party/terraform/tests/resource_network_services_gateway_test.go.erb
Comment (shuyama1) [inline mmv1/third_party/terraform/tests/resource_network_services_gateway_test.go.erb:276]: checked the debug logs for `TestAccNetworkServicesGateway_multipleSwpGatewaysDifferentSubnetwork` and `TestAccNetworkServicesGateway_multipleSwpGatewaysDifferentNetwork`. It seems that deletion works as intended.

**Review by shuyama1 — COMMENTED** [review]
Thanks for making the contribution!. I made a first pass and left some comments.

File: mmv1/products/networkservices/Gateway.yaml
Comment (felipegc) [reply mmv1/products/networkservices/Gateway.yaml]: The fields certificateUrls and gatewaySecurityPolicy will soon be updatable. I already created the task to revisit this once this is available and I will remove the immutable from them and enable the tests for them.

Done

File: mmv1/products/networkservices/Gateway.yaml
Comment (felipegc) [reply mmv1/products/networkservices/Gateway.yaml]: This field is not updatable so I am putting the immutable back.

Done

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (felipegc) [reply mmv1/templates/terraform/constants/network_services_gateway.go.erb:49]: Done

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (felipegc) [reply mmv1/templates/terraform/constants/network_services_gateway.go.erb]: Done

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (felipegc) [reply mmv1/templates/terraform/constants/network_services_gateway.go.erb]: Done

File: mmv1/templates/terraform/constants/network_services_gateway.go.erb
Comment (felipegc) [reply mmv1/templates/terraform/constants/network_services_gateway.go.erb]: Done

File: mmv1/products/networkservices/Gateway.yaml
Comment (felipegc) [reply mmv1/products/networkservices/Gateway.yaml]: Done

File: mmv1/third_party/terraform/tests/resource_network_services_gateway_test.go.erb
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_network_services_gateway_test.go.erb:277]: Great. Thanks.

**Comment by felipegc** [comment]
@shuyama1 Could you review again please? I managed to fix all issues you have raised. Thanks

File: mmv1/products/networkservices/Gateway.yaml
Comment (shuyama1) [inline mmv1/products/networkservices/Gateway.yaml:165]: what's the current behavior if users try to update the field? Will the API return errors in all cases? Or will it silently ignore the request in some cases? I understand that the field might need to set as immutable in the first place, as shown in the field description, but we should be extra cautious about adding `forcenew` to a field to ensure backward-compatibility, as now it will delete and recreate the resource.

File: mmv1/products/networkservices/Gateway.yaml
Comment (shuyama1) [inline mmv1/products/networkservices/Gateway.yaml:98]: Have you by chance tested this field if it works as intended when setting as `false`?

Comment (felipegc) [reply mmv1/products/networkservices/Gateway.yaml:98]: Yes. I generated the beta-provider and played with that. If it sets to false, it will leave the swg-autogen-router behind.

File: mmv1/products/networkservices/Gateway.yaml
Comment (felipegc) [reply mmv1/products/networkservices/Gateway.yaml:165]: This field supposed not to be updatable from the first place. The apis throws an error when trying to update only the scope. And by adding immutable:true it forces to recreate the resource.
I managed to test this scenario and worked as described.

**Comment by felipegc** [comment]
Hi @shuyama1 Did you have chance to take a look at the conversations? Please, close them if everything is clear.
Thanks

**Comment by shuyama1** [comment]
@felipegc Thanks for the ping. Running some local tests now. I'll give a final pass after and merge the PR if it's good to go or let you know if there's further change needed. Thanks!

**Review by shuyama1 — APPROVED** [review]
Thank you!

**Pull request merged**
